File: sdks/cpp/examples/3d_scene.cpp
/**
 * 3D Scene Demo
 * 
 * Demonstrates 3D rendering with PBR materials, lighting, and post-processing.
 * 
 * Build with: cmake --build build && ./build/examples/3d_scene
 */

#include <windjammer/windjammer.hpp>

using namespace windjammer;

int main() {
    App app;
    
    app.add_startup_system([]() {
        // Camera
        Camera3D{
            .position = Vec3{0.0f, 5.0f, 10.0f},
            .look_at = Vec3{0.0f, 0.0f, 0.0f},
            .fov = 60.0f
        };
        
        // Lights
        PointLight{ Vec3{5.0f, 5.0f, 5.0f}, Color{1.0f, 0.8f, 0.6f, 1.0f}, 2000.0f };
        PointLight{ Vec3{-5.0f, 5.0f, 5.0f}, Color{0.6f, 0.8f, 1.0f, 1.0f}, 1500.0f };
        PointLight{ Vec3{0.0f, 10.0f, -5.0f}, Color{1.0f, 1.0f, 1.0f, 1.0f}, 1000.0f };
        
        // Meshes with PBR materials
        Mesh::cube(1.0f).with_material(Material{
            .albedo = Color{0.8f, 0.2f, 0.2f, 1.0f},
            .metallic = 0.8f,
            .roughness = 0.2f,
            .emissive = Color{0.5f, 0.1f, 0.1f, 1.0f}
        });
        
        Mesh::sphere(1.0f, 32).with_material(Material{
            .albedo = Color{0.2f, 0.2f, 0.8f, 1.0f},
            .metallic = 0.5f,
            .roughness = 0.5f,
            .emissive = Color{0.1f, 0.1f, 0.5f, 1.0f}
        });
        
        Mesh::plane(10.0f).with_material(Material{
            .albedo = Color{0.3f, 0.3f, 0.3f, 1.0f},
            .metallic = 0.0f,
            .roughness = 0.9f
        });
        
        // Post-processing
        auto post = PostProcessing{};
        post.enable_hdr(true);
        post.set_bloom(BloomSettings{ .threshold = 1.0f, .intensity = 0.8f, .radius = 4.0f, .soft_knee = 0.5f });
        post.set_ssao(SSAOSettings{ .radius = 0.5f, .intensity = 1.5f, .bias = 0.025f, .samples = 16 });
        post.set_tone_mapping(ToneMappingMode::ACES, 1.2f);
        post.set_color_grading(ColorGrading{ .temperature = 0.1f, .tint = 0.0f, .saturation = 1.2f, .contrast = 1.1f });
    });
    
    app.add_system([](const Time& time) {
        // Rotate objects for dynamic lighting
    });
    
    app.run();
    return 0;
}

File: sdks/ruby/examples/3d_scene.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# 3D Scene Demo
#
# Demonstrates 3D rendering with PBR materials, lighting, and post-processing.
#
# Run with: ruby examples/3d_scene.rb

require 'windjammer'

app = Windjammer::App.new

app.add_startup_system do
  # Camera
  Windjammer::Camera3D.new(
    position: Windjammer::Vec3.new(0, 5, 10),
    look_at: Windjammer::Vec3.new(0, 0, 0),
    fov: 60.0
  )
  
  # Lights
  Windjammer::PointLight.new(position: Windjammer::Vec3.new(5, 5, 5), color: Windjammer::Color.new(1.0, 0.8, 0.6, 1.0), intensity: 2000.0)
  Windjammer::PointLight.new(position: Windjammer::Vec3.new(-5, 5, 5), color: Windjammer::Color.new(0.6, 0.8, 1.0, 1.0), intensity: 1500.0)
  Windjammer::PointLight.new(position: Windjammer::Vec3.new(0, 10, -5), color: Windjammer::Color.new(1, 1, 1, 1), intensity: 1000.0)
  
  # Meshes with PBR materials
  Windjammer::Mesh.cube(size: 1.0).with_material(Windjammer::Material.new(
    albedo: Windjammer::Color.new(0.8, 0.2, 0.2, 1.0),
    metallic: 0.8,
    roughness: 0.2,
    emissive: Windjammer::Color.new(0.5, 0.1, 0.1, 1.0)
  ))
  
  Windjammer::Mesh.sphere(radius: 1.0, subdivisions: 32).with_material(Windjammer::Material.new(
    albedo: Windjammer::Color.new(0.2, 0.2, 0.8, 1.0),
    metallic: 0.5,
    roughness: 0.5,
    emissive: Windjammer::Color.new(0.1, 0.1, 0.5, 1.0)
  ))
  
  Windjammer::Mesh.plane(size: 10.0).with_material(Windjammer::Material.new(
    albedo: Windjammer::Color.new(0.3, 0.3, 0.3, 1.0),
    metallic: 0.0,
    roughness: 0.9
  ))
  
  # Post-processing
  post = Windjammer::PostProcessing.new
  post.enable_hdr(true)
  post.set_bloom(Windjammer::BloomSettings.new(threshold: 1.0, intensity: 0.8, radius: 4.0, soft_knee: 0.5))
  post.set_ssao(Windjammer::SSAOSettings.new(radius: 0.5, intensity: 1.5, bias: 0.025, samples: 16))
  post.set_tone_mapping(Windjammer::ToneMappingMode::ACES, 1.2)
  post.set_color_grading(Windjammer::ColorGrading.new(temperature: 0.1, tint: 0.0, saturation: 1.2, contrast: 1.1))
end

app.add_system do |time|
  # Rotate objects for dynamic lighting
end

app.run
